temporalio/common.py

```diff
@@ -175,6 +184,334 @@ def __setstate__(self, state: object) -> None:
+        elif isinstance(vals[0], bool):
+            return temporalio.common.SearchAttributeKey.for_bool(name)
+        elif isinstance(vals[0], datetime):
+            return temporalio.common.SearchAttributeKey.for_datetime(name)
+        return None
+
+
+@dataclass(frozen=True)
+class _SearchAttributeKey(SearchAttributeKey[SearchAttributeValueType]):
+    _name: str
+    _indexed_value_type: SearchAttributeIndexedValueType
+    # No supported way in Python to derive this, so we're setting manually
+    _value_type: Type[SearchAttributeValueType]
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def indexed_value_type(self) -> SearchAttributeIndexedValueType:
+        return self._indexed_value_type
+
+    @property
+    def value_type(self) -> Type[SearchAttributeValueType]:
+        return self._value_type
+
+
+class SearchAttributePair(NamedTuple, Generic[SearchAttributeValueType]):
+    """A named tuple representing a key/value search attribute pair."""
+
+    key: SearchAttributeKey[SearchAttributeValueType]
+    value: SearchAttributeValueType
+
+
+class SearchAttributeUpdate(ABC, Generic[SearchAttributeValueType]):
+    """Representation of a search attribute update."""
+
+    @property
+    @abstractmethod
+    def key(self) -> SearchAttributeKey[SearchAttributeValueType]:
+        """Key that is being set."""
+        ...
+
+    @property
+    @abstractmethod
+    def value(self) -> Optional[SearchAttributeValueType]:
+        """Value that is being set or ``None`` if being unset."""
+        ...
+
+
+@dataclass(frozen=True)
+class _SearchAttributeUpdate(SearchAttributeUpdate[SearchAttributeValueType]):
+    _key: SearchAttributeKey[SearchAttributeValueType]
+    _value: Optional[SearchAttributeValueType]
+
+    @property
+    def key(self) -> SearchAttributeKey[SearchAttributeValueType]:
+        return self._key
+
+    @property
+    def value(self) -> Optional[SearchAttributeValueType]:
+        return self._value
+
+
+@dataclass(frozen=True)
+class TypedSearchAttributes(Collection[SearchAttributePair]):
+    """Collection of typed search attributes.
+
+    This is represented as an immutable collection of
+    :py:class:`SearchAttributePair`. This can be created passing a sequence of
+    pairs to the constructor.
+    """
+
+    search_attributes: Sequence[SearchAttributePair]
+    """Underlying sequence of search attribute pairs. Do not mutate this, only
+    create new ``TypedSearchAttribute`` instances.
+
+    These are sorted by key name during construction.
+    """
+
+    empty: ClassVar[TypedSearchAttributes]
+    """Class variable representing an empty set of attributes."""
+
+    def __post_init__(self):
+        """Post-init initialization."""
+        # Sort
+        object.__setattr__(
+            self,
+            "search_attributes",
+            sorted(self.search_attributes, key=lambda pair: pair.key.name),
```

Comment:
Why is this important?

Reply:
Predictability for in-workflow and out-of-workflow use for those ranging. For me personally, it was so I could use `==` in my tests.

Reply:
Hrmm, yeah I should fail on duplicates given I have a `updated` helper

Reply:
Ok, disallowed duplicates (post sort, I still want sort)